Alchemy 6 multi search

kulturbande opened the pull request.

tvdeyen commented on alchemy-pg_search.gemspec:
```diff
@@ -16,7 +16,7 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
 
-  spec.add_runtime_dependency "alchemy_cms", [">= 5.0", "< 6"]
+  spec.add_runtime_dependency "alchemy_cms", [">= 5.0", "< 7"]
```
Muss angehoben werden

```suggestion
  spec.add_runtime_dependency "alchemy_cms", [">= 6.0", "< 7"]
```

tvdeyen commented on app/jobs/alchemy/pg_search/index_page_job.rb:
```diff
@@ -0,0 +1,10 @@
+  module PgSearch
+    class IndexPageJob < BaseJob
+      def perform(page)
+        Search::index_page page
```
Geht so zwar auch, ist aber eher ungewöhnlicher Ruby Syntax

```suggestion
        Search.index_page(page)
```

tvdeyen commented on app/views/alchemy/search/_results.html.erb:
```diff
@@ -10,12 +10,8 @@
-          options: options,
-          page: page
-        ) %>
+      <% @search_results.each do |page_result| %>
+        <%= render('alchemy/search/result', result: page_result) %>
       <% end %>
```
Sollte sogar so gehen

```suggestion
      <%= render(partial: 'alchemy/search/result', collection: @search_results) %>
```

tvdeyen commented on db/migrate/20220714183808_create_pg_search_documents.rb:
```diff
@@ -0,0 +1,18 @@
+class CreatePgSearchDocuments < ActiveRecord::Migration[6.0]
```
Das sollten wir nicht in diesem Gem machen, da es sonst sein kann, dass wir über die Zeit mit pg_search divergieren. Statt dessen sollten wir in der README erwähnen den Generator von pg_search auszuführen. 
https://github.com/Casecommons/pg_search#setup

kulturbande replied:
Bin ich mir nicht sicher, da die Tabelle auch bearbeit wird und die page_id mit integriert wird. Wenn sich die Migrationen anpassen, müsste das Gem eh angepasst werden.

tvdeyen replied:
Ja, das ist in der Tat wichtig. Wir sollten trotzdem die beiden Migrationen trennen (also die PgSearch Migration aus dem Gem kommen lassen). Und in unserem Gem die `page_id` in den Table hinzufügen. Die Trennung ist wichtig im Kontext einer Nutzung in einer Rails app, die wir nicht kennen.

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -19,5 +15,13 @@ def self.is_searchable_essence?(essence_type)
     end
+    
+    def self.search(query, ability: nil)
+      Search.search query, ability: ability
```
Ruby braucht zwar keine Klammern, aber es liest sich doch schöner:

```suggestion
      Search.search(query, ability: ability)
```

tvdeyen commented on lib/alchemy/pg_search/content_extension.rb:
```diff
@@ -10,6 +10,8 @@ def new(attributes)
 
   module InstanceMethods
+    ##
+    # @deprecated
```
Wolltest Du das auch noch löschen?

tvdeyen commented on lib/alchemy/pg_search/controller_methods.rb:
```diff
@@ -49,31 +49,24 @@ def perform_search
-        else
-          pages.not_restricted.full_text_search(params[:query])
-        end
+        Alchemy::PgSearch.search params[:query], ability: current_ability
```
Je mehr man mit anderen Programmiersprachen zu tun hat, desto mehr mag man es doch mit Klammern:

```suggestion
        Alchemy::PgSearch.search(params[:query], ability: current_ability)
```

tvdeyen commented on lib/alchemy/pg_search/page_extension.rb:
```diff
@@ -1,54 +1,25 @@
-      )
-    end
+  def remove_unpublished_pages
+    Alchemy::PgSearch::Search.remove_page self unless searchable?
```
Kudos an den Ruby parser, dass der das auch ohne Klammern schnallt

```suggestion
    Alchemy::PgSearch::Search.remove_page(self) unless searchable?
```

tvdeyen commented on lib/alchemy/pg_search/page_extension.rb:
```diff
@@ -1,54 +1,25 @@
 # Enable Postgresql full text indexing.
 #
-module Alchemy::PgSearch::PageExtension
```
Eigentlich ist das der bessere Syntax, statt dem `class_eval`.

tvdeyen commented on lib/alchemy/pg_search/page_extension.rb:
```diff
@@ -1,54 +1,25 @@
-      all_elements.full_text_search(query)
-    end
+  def searchable?
+    public? && !layoutpage?
```
Geht auch positiv und ein µ schneller ;)

```suggestion
    contentpage? && public?
```

tvdeyen commented on lib/alchemy/pg_search/pg_search_document_extension.rb:
```diff
@@ -0,0 +1,14 @@
+PgSearch::Document.class_eval do
```
Würde das auch gerne eher als `Module#prepend` schreiben

tvdeyen commented on lib/alchemy/pg_search/search.rb:
```diff
@@ -0,0 +1,53 @@
+      ##
+      # index all supported Alchemy models
+      def self.rebuild
+        [Alchemy::Page, Alchemy::EssenceText, Alchemy::EssenceRichtext, Alchemy::EssencePicture].each do |model|
```
Das wollen wir später idealerweise konfigurierbar machen und wir müssen für volle Alchemy 6 Kompatibilität auch noch die Ingredients einplanen.

tvdeyen commented on lib/alchemy/pg_search/search.rb:
```diff
@@ -0,0 +1,53 @@
+      # search for page results
+      #
+      # @param query [string]
+      def self.search(query, ability: nil)
```
Wir sollten die `ability` als Argument auch noch dokumentieren.

tvdeyen reviewed (commented):
Sehr gute Arbeit. Ich habe noch eine Menge kleinerer Kommentare, die aber eher Nichtigkeiten sind. Wir sollten den PR noch so aufräumen, das er mehrere zusammenhängende Commits hat (zB. das Alchemy Update in der Dummy app, das hinzufügen des multisearchs und dann die Entfernung des alten Codes, oder so ähnlich)

tvdeyen commented on lib/alchemy/pg_search/search.rb:
```diff
@@ -0,0 +1,59 @@
+        remove_page page
+
+        page.update_pg_search_document
+        page.all_elements.includes(contents: :essence).each do |element|
```
Eventuell wollen noch eine Optimierung für sehr große elements table machen

```suggestion
        page.all_elements.includes(contents: :essence).find_each do |element|
```

Das lädt die Elements in packen zu tausend anstatt alle auf einmal

tvdeyen replied:
Und wir wollen auch die `ingredients` eager loaden.

```suggestion
        page.all_elements.includes(contents: :essence, :ingredients).find_each do |element|
```

kulturbande committed:
Update Dummy App to Alchemy 6

- updated dummy app from Alchemy 5.1 to Alchemy 6
- added also the new migrations to dummy app
- added IDE config folders to gitignore

kulturbande committed:
Enable pg_search multisearch

- added install script which enabled pg_search multisearch
- multisearch is much faster and easier to extend

kulturbande committed:
Moved all extensions to a separate folder

- moved the extensions to app/extension - folder to improve readability
- extended also the pg_seearch_document - model to use alchemy page reference
- extended essences as well as ingredients for text, richtext, and picture
- marked the gem to be only compatible with Alchemy 6.x (Ingredients were introduced with Alchemy 6)

kulturbande committed:
Introduced new search module

- a single place to search and reindex pages or all supported models at once
- changed controller and views to use the new queried results
- add index job as an hook to reindex the page if something changed
- updated the fulltext_search_feature to use the page_version mechanic of Alchemy 6

tvdeyen reviewed (approved):
Very nice work! Thank you
